File: src/taller/taller3/DiscountRechargeableSubwayCard.java
package taller.taller3;

public class DiscountRechargeableSubwayCard extends RechargeableSubwayCard {
    private double discountPercentaje;

    DiscountRechargeableSubwayCard(SubwayCentral sc, double dp) {
        super(sc);
        discountPercentaje = dp;
    }

    @Override
    protected double rideCost() {
        return super.rideCost() * discountPercentaje;
    }
}

File: POO-Ruby/guias/guia8/ej6/discount_rechargeable_subway_card.rb
# frozen_string_literal: true
require_relative 'rechargeable_subway_card'

class DiscountRechargeableSubwayCard < RechargeableSubwayCard
  def initialize(subway_central, discount_percentage)
    super(subway_central)
    raise 'invalid discount' unless discount_percentage.is_a?(Numeric) && discount_percentage.positive?
    @dp = discount_percentage
  end

  private def ride_cost
    super * (1 - @dp)
  end
end
